p2p/net/swarm/dial_sync.go

```diff
@@ -101,9 +111,10 @@ func (ds *dialSync) Dial(ctx context.Context, p peer.ID) (*Conn, error) {
 		defer ds.mutex.Unlock()
 		ad.refCnt--
 		if ad.refCnt == 0 {
-			ad.close()
+			ad.close(err)
 			delete(ds.dials, p)
 		}
 	}()
-	return ad.dial(ctx)
+	conn, err := ad.dial(ctx) // updated err is used in defered func
```

Comment:
Are you sure this works? Isn't `:=` creating a new `err` here?

Why do we have this `defer` anyway, there's only a single return from this function.

Reply:
err is declared in the beginning so this will assign to it. 

Anyway, I will remove the defer. That is confusing.

Reply:
removed it.